compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp

```diff
@@ -0,0 +1,233 @@
+
+  IRRewriter rewriter(context);
+  FailureOr<TilingInfo> tilingInfo =
+      getTiledAndDistributionInfo(rewriter, computeOps);
+  if (failed(tilingInfo)) {
+    return signalPassFailure();
+  }
+  auto tilableOp = dyn_cast_or_null<TilingInterface>(tilingInfo->tilableOp);
+  if (!tilableOp) {
+    // Did not find a tileable op. So do nothing.
+    return;
+  }
+
+  scf::SCFTilingOptions tilingOptions;
+  tilingOptions.setTileSizes(tilingInfo->tileSizes);
+  tilingOptions.setInterchange(tilingInfo->interchange);
+  tilingOptions.setLoopType(scf::SCFTilingOptions::LoopType::ForallOp);
+  SmallVector<Attribute> deviceMappingAttribute =
+      getMapping(context, tilingInfo->tileSizes);
+  if (failed(IREE::Codegen::WorkgroupMappingAttr::verifyAttrList(
+          context, ::mlir::detail::getDefaultDiagnosticEmitFn(funcOp.getLoc()),
+          deviceMappingAttribute))) {
+    return signalPassFailure();
+  }
+  tilingOptions.setMapping(deviceMappingAttribute);
+
+  scf::SCFTileAndFuseOptions tileAndFuseOptions;
+  tileAndFuseOptions.setTilingOptions(tilingOptions);
+  // TODO: For now use the default tile and fuse control function. That needs
+  // to be modified to allow for returning the values of the producer when
+  // needed.
+  rewriter.setInsertionPoint(tilableOp);
+
+  // If the `tilableOp` is a `memref` op, then just tile the operation.
+  SmallVector<LoopLikeOpInterface> tilingLoops;
+  if (tilableOp->getNumResults() == 0) {
+    FailureOr<scf::SCFTilingResult> tilingResult =
+        scf::tileUsingSCF(rewriter, tilableOp, tilingOptions);
+    if (failed(tilingResult)) {
+      funcOp.emitOpError("tiling failed");
+      return signalPassFailure();
+    }
+    rewriter.eraseOp(tilableOp);
+    std::swap(tilingResult->loops, tilingLoops);
+  } else {
+    FailureOr<scf::SCFTileAndFuseResult> tileAndFuseResult =
+        scf::tileConsumerAndFuseProducersUsingSCF(rewriter, tilableOp,
+                                                  tileAndFuseOptions);
+    if (failed(tileAndFuseResult)) {
+      funcOp.emitOpError("tile and fuse greedily failed");
+      return signalPassFailure();
+    }
+    for (auto [origValue, replacement] : tileAndFuseResult->replacements) {
+      rewriter.replaceAllUsesWith(origValue, replacement);
+    }
+    std::swap(tileAndFuseResult->loops, tilingLoops);
+  }
+
+  // Cleanup patterns for tile and distribute
+  {
+    SmallVector<int64_t> staticNumIters;
+    if (!tilingLoops.empty()) {
+      assert(tilingLoops.size() == 1 &&
+             "expected only a single scf.forall loop");
+      auto forallLoop = cast<scf::ForallOp>(tilingLoops.front());
+      staticNumIters = getStaticNumWorkgroups(rewriter, forallLoop);
+    }
+    RewritePatternSet patterns(context);
+    linalg::populateLinalgTilingCanonicalizationPatterns(patterns);
+    tensor::populateFoldTensorEmptyPatterns(patterns);
+    populateFoldAffineMinInDistributedLoopsPatterns(patterns, staticNumIters);
```

Comment:
Looking at the `populateFoldAffineMinInDistributedLoops` pattern, it looks like it only applies to distributed `scf.for` ops. Can we drop this and `getStaticNumWorkgroups`?

Reply:
I went back and forth on this, and I felt I might eventually need to do this so did it now. But you convinced me to drop it.